service-base/src/main/java/org/eclipse/hono/service/metric/MicrometerBasedMetrics.java

```diff
@@ -106,29 +119,80 @@ public long getNumberOfConnections() {
     }
 
     @Override
-    public final void incrementProcessedMessages(final String type, final String tenantId) {
-
-        Objects.requireNonNull(type);
-        Objects.requireNonNull(tenantId);
-        this.registry.counter(METER_MESSAGES_PROCESSED,
-                Tags.of(MetricsTags.TAG_TENANT, tenantId).and(MetricsTags.TAG_TYPE, type))
-                .increment();
+    public Sample startTimer() {
+        return Timer.start(registry);
+    }
 
+    @Override
+    public final void reportTelemetry(
+            final EndpointType type,
+            final String tenantId,
+            final ProcessingOutcome outcome,
+            final QoS qos,
+            final Sample timer) {
+
+        reportTelemetry(type, tenantId, outcome, qos, TtdStatus.NONE, timer);
     }
 
     @Override
-    public final void incrementUndeliverableMessages(final String type, final String tenantId) {
+    public final void reportTelemetry(
+            final EndpointType type,
+            final String tenantId,
+            final ProcessingOutcome outcome,
+            final QoS qos,
+            final TtdStatus ttdStatus,
+            final Sample timer) {
 
         Objects.requireNonNull(type);
         Objects.requireNonNull(tenantId);
-        this.registry.counter(METER_MESSAGES_UNDELIVERABLE,
-                Tags.of(MetricsTags.TAG_TENANT, tenantId).and(MetricsTags.TAG_TYPE, type))
-                .increment();
+        Objects.requireNonNull(outcome);
+        Objects.requireNonNull(ttdStatus);
+        Objects.requireNonNull(timer);
+
+        if (!EndpointType.TELEMETRY.equals(type) && !EndpointType.EVENT.equals(type)) {
+            throw new IllegalArgumentException("invalid type, must be either telemetry or event");
+        }
 
+        timer.stop(this.registry.timer(
+                METER_MESSAGES_RECEIVED,
+                Tags.of(MetricsTags.getTypeTag(type))
+                    .and(MetricsTags.getTenantTag(tenantId))
```

Comment:
I don't think calling "and" with `null` will work. Looking into the callstack of `Tags`, this seems to boil down to:

~~~java
Map<String, Tag> merged = new LinkedHashMap<>(this.tags);
tags.forEach(tag -> merged.put(tag.getKey(), tag));
return new Tags(merged);
~~~

Which should throw a NPE in the case of a `null` element.

Reply:
Tags.and(Tag ...):
```
    /**
     * Return a new {@link Tags} instance by merging this collection and the specific tags.
     *
     * @param tags the tags to add
     * @return a new {@link Tags} instance
     */
    public Tags and(@Nullable Tag... tags) {
        if (tags == null || tags.length == 0) {
            return this;
        }
        return and(Arrays.asList(tags));
    }
```
this looks as if it would accept a `null` arg, doesn't it?

Reply:
Yes, I ran the same test :-) With the same result (obviously).

Reply:
@ctron  Based on this discussion I wonder if we also should not create a *ttd* tag for `TtdStatus.NONE`. WDYT?

Reply:
@ctron Given that we are reporting every single message being sent, I think that we should also probably not instantiate new Tag instances for each message. Instead, we could create the Tag instances as part of the enums and re-use the instances, thus reducing GC pressure. Do you think that would be worth the effort?

Reply:
I agree, that might be a good idea.